scripts/core/editor3/helpers/resizeDraftSelection.js

```diff
@@ -0,0 +1,164 @@
+            blockKey: boundaryBlockKey,
+            offset: boundaryOffset
+        };
+    }
+
+    // check if resizing can be resolved in the current block
+
+    const currentBlock = contentState.getBlockForKey(currentBlockKey);
+
+    if (currentOffset + expandBy <= currentBlock.getLength()) {
+        return {
+            blockKey: currentBlockKey,
+            offset: currentOffset + expandBy
+        };
+    }
+
+    // jump to the start of the block after
+
+    const blockAfter = contentState.getBlockAfter(currentBlockKey);
+
+    if (blockAfter === null || blockAfter === undefined || limitedToSingleBlock) {
+        return {
+            blockKey: currentBlockKey,
+            offset: currentBlock.getLength()
+        };
+    }
+
+    const nextStretchRight = (currentOffset + expandBy) - currentBlock.getLength();
+    const nextOffset = 0;
+    const nextBlockKey = blockAfter.getKey();
+
+    return expandRightSingleCursor(
+        nextStretchRight,
+        boundaryBlockKey,
+        boundaryOffset,
+        nextBlockKey,
+        nextOffset,
+        limitedToSingleBlock,
+        contentState
+    );
+}
+
+function expandLeftSingleCursor(
+    stretchLeft,
+    boundaryBlockKey,
+    boundaryOffset,
+    currentBlockKey,
+    currentOffset,
+    limitedToSingleBlock,
+    contentState
+) {
+    const boundaryViolated = currentBlockKey === boundaryBlockKey && currentOffset - stretchLeft <= boundaryOffset;
+
+    if (boundaryViolated) {
+        return {
+            blockKey: boundaryBlockKey,
+            offset: boundaryOffset
+        };
+    }
+
+    // check if resizing can be resolved in the current block
+
+    if (currentOffset - stretchLeft >= 0) {
+        return {
+            blockKey: currentBlockKey,
+            offset: currentOffset - stretchLeft
+        };
+    }
+
+    // jump to the end of the block before
+
+    const blockBefore = contentState.getBlockBefore(currentBlockKey);
+
+    if (blockBefore === null || blockBefore === undefined || limitedToSingleBlock) {
+        return {
+            blockKey: currentBlockKey,
+            offset: 0
+        };
+    }
+
+    const nextStretchLeft = stretchLeft - currentOffset;
+    const nextOffset = blockBefore.getLength();
+    const nextBlockKey = blockBefore.getKey();
+
+    return expandLeftSingleCursor(
+        nextStretchLeft,
+        boundaryBlockKey,
+        boundaryOffset,
+        nextBlockKey,
+        nextOffset,
+        limitedToSingleBlock,
+        contentState
+    );
+}
+
+export function resizeDraftSelection(stretchLeft, stretchRight, selection, editorState, limitedToSingleBlock = false) {
+    const contentState = editorState.getCurrentContent();
+    const firstBlock = contentState.getFirstBlock();
+    const lastBlock = contentState.getLastBlock();
+
+    // shrinking is performed by extending in the opposite direction
+    const shrinkLeft = stretchLeft < 0;
+    const nextLeft = shrinkLeft
+        ? expandRightSingleCursor(
```

Comment:
All my tests failed when I tried to use `getBlockAndOffset`. I've tested `getBlockAndOffset` manually and it doesn't seem to work correctly.

![selection](https://user-images.githubusercontent.com/2076953/37512349-95588aa6-290a-11e8-9d59-0f36e05db533.png)
`getBlockAndOffset(contentState, currentSelection, -5)` I expected `offset` to be the same as the `anchorKey`(27) since, I moved the selection to left by it's length. I received `22` instead.

Reply:
The function moves the beginning of the selection, so the offset 22 is correct.
Even if there is a bug in a code I think it is not a good reason to add a completely new code and duplicate the code. Instead it will be better to refactor the code even if this can be the completely deletion of the old code. It is important to eliminate the code duplication as much as possible.

Reply:
You're right about the offset, I don't know why I expected it to start from focus offset.

I also agree avoiding code duplication and propose that I remove `getBlockAndOffset` and replace it with `expandLeftSingleCursor` in all places it's used. Is that alright?

I believe I found a real bug this time.When calling `getBlockAndOffset`, with a selection start on 0, and are trying to go -1, the selection jumps to the previous block correctly, but instead of stopping at the last position, which would be 30 in this case, it goes to 28. Do you think it would cause problems in the code using `getBlockAndOffset` if I replaced it to `expandLeftSingleCursor` ?

Reply:
Forgot to mention the reasons I prefer `expandLeftSingleCursor`. That's mostly a support for expanding right, not only to the left without needing to know content length. It's already implemented correctly and has tests. It also supports out of range values and prevents from producing invalid selection objects. I think it's not possible to refactor `getBlockAndOffset` to support right direction without splitting it into two functions anyway.

Reply:
Indeed there was a bug on getBlockAndOffset. Thank you for founding it.
Looks like the right offset should be 29 because you don't have to count when you move from the first position of the block to the last position of the previous block. The last offset in the block is just after the last character.
Here is a PR (https://github.com/superdesk/superdesk-client-core/pull/2154) that fixes the bug and add support for moving by positive and negative offsets starting from the start or from the end of the selection.
So I don't see any improvement in replacing one existing simple function with two new more large, complicated(more parameters, recursive function) and inflexible functions(the large if expressions). The only improvement are the tests and those tests are not lost because are at resizeDraftSelection level.